tools/toitlsp/lsp/compiler/compiler_fs_protocol.go

```diff
@@ -0,0 +1,244 @@
+// Copyright (C) 2022 Toitware ApS.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; version
+// 2.1 only.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// The license can be found in the file `LICENSE` in the top level
+// directory of this repository.
+
+package compiler
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"sync"
+
+	cpath "github.com/toitware/toit.git/toitlsp/lsp/compiler/path"
+	"go.uber.org/zap"
+)
+
+type FileSystem interface {
+	Read(path string) (File, error)
+	ListDirectory(path string) ([]string, error)
+	PackageCachePaths() ([]string, error)
+}
+
+type File struct {
+	Path        string
+	Exists      bool
+	IsRegular   bool
+	IsDirectory bool
+	Content     []byte
+}
+
+type CompilerFSProtocol struct {
+	// l handles the fields listener and closeCh
+	l sync.Mutex
+
+	fileCache      *fileCache
+	directoryCache *directoryCache
+	logger         *zap.Logger
+	sdkPath        string
+	fs             FileSystem
+
+	servedSdkPath           *string
+	servedPackageCachePaths []string
+}
+
+func NewCompilerFSProtocol(fs FileSystem, logger *zap.Logger, SDKPath string) *CompilerFSProtocol {
+	return &CompilerFSProtocol{
+		fileCache:      newFileCache(fs),
+		directoryCache: newDirectoryCache(fs),
+		logger:         logger.Named("compiler_protocol"),
+		sdkPath:        SDKPath,
+		fs:             fs,
+	}
+}
+
+func (cp *CompilerFSProtocol) HandleConn(conn io.ReadWriter) {
```

Comment:
The code here has just been moved from file_server.go
The only lines changed are at the very top (`scanner` and `w`)